### PR TITLE
ISPN-5228 ISPN-5229 Cleanup of redundant code

### PR DESCRIPTION
**Pull request by rvansa**
https://issues.jboss.org/browse/ISPN-5228
https://issues.jboss.org/browse/ISPN-5229

Implementation of many visitX commands was almost copy-paste - this PR unites them.
Main focus is on *LockingInterceptor, and a bit on CacheLoaderInterceptor.

### CODE REVIEW

**Comment by tristantarrant** [comment]
Can you enclose this in a Jira please ?

**Comment by rvansa** [comment]
OK, updated. It seemed to me that clean-up PR won't need JIRA entries, but if that's preferred, I don't have objections.

File: core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java [hunk 1]
```diff
@@ -93,46 +87,22 @@ protected void injectDependencies(PersistenceManager clm, EntryFactory entryFact
...
-      return invokeNextInterceptor(ctx, command);
+      return visitDataCommand(ctx, command);
    }
 
    @Override
    public Object visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand command) throws Throwable {
-      if (enabled) {
-         Object key;
-         if ((key = command.getKey()) != null) {
-            loadIfNeeded(ctx, key, command);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: Nice we weren't incrementing the count properly on a miss before.

Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: Actually looking closer it seems then that loadIfNeededAndUpdateStats should be merged into loadIfNeeded so we won't have this issue in the future.  This would prevent ActivationInterceptor visitPutMapCommand from missing a count as well then.

File: core/src/main/java/org/infinispan/interceptors/locking/AbstractTxLockingInterceptor.java [hunk 2]
```diff
@@ -56,20 +57,18 @@ public Object visitRollbackCommand(TxInvocationContext ctx, RollbackCommand comm
...
+      if (command.hasFlag(Flag.PUT_FOR_EXTERNAL_READ)) {
+         // Cache.putForExternalRead() is non-transactional
+         return visitNonTxDataWriteCommand(ctx, command);
       }
+      return visitDataWriteCommand(ctx, command);
    }
 
    @Override
-   public Object visitGetCacheEntryCommand(InvocationContext ctx, GetCacheEntryCommand command) throws Throwable {
+   protected Object visitDataReadCommand(InvocationContext ctx, DataCommand command) throws Throwable {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/locking/AbstractTxLockingInterceptor.java]: It seems this is only used in OptimisticLockingInterceptor as Pessimistic doesn't call super so I would say to just move this method into Optimistic directly.

File: core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java (hunk 1 above)
Comment (rvansa) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: I really don't know the ratio why InvalidateCommand is not increasing the missed counters, and the goal was not really to fix the behaviour but rather reduce code complexity. And we need different method for Invalidate and PutMap anyway.

File: core/src/main/java/org/infinispan/interceptors/locking/AbstractTxLockingInterceptor.java (hunk 2 above)
Comment (rvansa) [inline core/src/main/java/org/infinispan/interceptors/locking/AbstractTxLockingInterceptor.java]: Makes sense, I'll do that.

File: core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java (hunk 1 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: True, I overlooked Invalidate, that shouldn't increment IMO.  In regards to the fixes I was more saying we fixed Put and Apply delta though which looks like an oversight before.

Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: Although I can see the opposite: People could think that put and delta write wouldn't affect the load count.  But if that is the case I would say that remove shouldn't be incrementing either, but it was before.  @danberindei @pruivo  What do you guys think ?

Comment (pruivo) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: IMO we should have this:
* the put should increment if miss because we return the previous value
* the InvalidateCommand (and subclasses) should not increment (internal use on L1 and state transfer)
* the PutMapCommand should not increment either (we don't need the previous values. actually, it does not need to interact with the cache loader)
* the delta is more complicated but I think it can count the misses. in the end, we need to load the object to apply the deltas.

Comment (rvansa) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: I have updated the PR, keeping only loadIfNeeded and doing the counter increments directly there.
After I've given it more thoughts, I think that it shouldn't matter what command triggered the operation - the counters should increment once we've physically touched the cache loader, so that's what the code should be doing now. The counters here are to let user see that application slowdown is there due to many loads/misses - what caused them is another issue.

Comment (pruivo) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: Agree with @rvansa, but I think the Invalidate should not load anything from the CacheLoader. It should remove the entry from memory and from private CacheStore (if not L1 invalidate). Anyway, this is outside the scope of this PR :)

Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: I agree as well (was writing up something else but this is short and sweet)

Also the InvalidateL1Command shouldn't exist in the store since we don't put it there on initial insert and they aren't passivated as of https://github.com/infinispan/infinispan/pull/2824

Comment (rvansa) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: @pruivo You need to send removal notifications, and for those you need the value - if it's not in memory it needs to be loaded.

Comment (danberindei) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: @rvansa yeah, it would be nice if we could do that only if we actually have a listener registered, but that's definitely outside the scope of this PR.

**Comment by galderz** [comment]
Integrated

File: core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java (hunk 1 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java]: I created https://issues.jboss.org/browse/ISPN-5296 so we can revisit this at some point.
